cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/LeaderReplicaDistributionGoal.java

```diff
@@ -294,16 +299,16 @@ private boolean rebalanceByMovingLeaderReplicasIn(Broker broker,
     List<Broker> candidateBrokers = Collections.singletonList(broker);
     Set<String> excludedTopics = optimizationOptions.excludedTopics();
     boolean onlyMoveImmigrantReplicas = optimizationOptions.onlyMoveImmigrantReplicas();
+    String replicaSortName = replicaSortName(this, false, true);
+    new SortedReplicasHelper().addSelectionFunc(ReplicaSortFunctionFactory.selectReplicasNotFromExcludedTopics(excludedTopics))
+                              .addSelectionFunc(ReplicaSortFunctionFactory.selectLeaders())
+                              .maybeAddSelectionFunc(ReplicaSortFunctionFactory.selectImmigrants(),
+                                                     !clusterModel.brokenBrokers().isEmpty() || onlyMoveImmigrantReplicas)
+                              .trackSortedReplicasFor(replicaSortName, clusterModel);
```

Comment:
Looks like we are potentially re-filtering (i.e. re-selecting) replicas for the whole cluster every time `rebalanceByMovingLeaderReplicasIn` is called. This seems expensive and maybe unnecessary. I am wondering if we can improve the performance here a little -- e.g. via calling `trackSortedReplicasFor` for the whole cluster just once during the initialization as we do in other goals?

Reply:
the sorted replica is created lazily, in most case we just do that for one or two broker and in each iteration it is probably different broker. Therefore doing in `initialization` probably would not save much. The main reason I do not do it in initialization code is that in `rebalanceByMovingReplicasOut` we also create sorted replica for some brokers, there will be  `replicaSortName` conflict for different sorted replica set for the same broker.